ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationsReOrgManager.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition;
+
+import java.util.NavigableMap;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.blocks.BeaconBlock;
+import tech.pegasys.teku.datastructures.blocks.BeaconBlockBody;
+import tech.pegasys.teku.datastructures.operations.AttesterSlashing;
+import tech.pegasys.teku.datastructures.operations.ProposerSlashing;
+import tech.pegasys.teku.datastructures.operations.SignedVoluntaryExit;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.statetransition.attestation.AggregatingAttestationPool;
+import tech.pegasys.teku.statetransition.attestation.AttestationManager;
+import tech.pegasys.teku.storage.api.ReorgEventChannel;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+public class OperationsReOrgManager implements ReorgEventChannel {
+  private static final Logger LOG = LogManager.getLogger();
+
+  final OperationPool<SignedVoluntaryExit> exitPool;
+  final OperationPool<ProposerSlashing> proposerSlashingPool;
+  final OperationPool<AttesterSlashing> attesterSlashingPool;
+  final AttestationManager attestationManager;
+  final AggregatingAttestationPool attestationPool;
+  final RecentChainData recentChainData;
+
+  public OperationsReOrgManager(
+      OperationPool<ProposerSlashing> proposerSlashingPool,
+      OperationPool<AttesterSlashing> attesterSlashingPool,
+      OperationPool<SignedVoluntaryExit> exitPool,
+      AggregatingAttestationPool attestationPool,
+      AttestationManager attestationManager,
+      RecentChainData recentChainData) {
+    this.exitPool = exitPool;
+    this.proposerSlashingPool = proposerSlashingPool;
+    this.attesterSlashingPool = attesterSlashingPool;
+    this.attestationManager = attestationManager;
+    this.attestationPool = attestationPool;
+    this.recentChainData = recentChainData;
+  }
+
+  @Override
+  public void reorgOccurred(
```

Comment:
This method is too big overall. Probably worth splitting in two - process nonCanonical and then process canonical blocks.  And then probably a few other helper methods to make the optional/future chainings not wind up so deeply nested and hard to follow.